File: 1089-duplicate-zeros/1089-duplicate-zeros.cpp
class Solution {
public:
    void duplicateZeros(vector<int>& arr) {
        int zeroCount = 0;
        int lastIdx = arr.size() - 1;

        for(int i = 0; i <= lastIdx - zeroCount; i++) {
            if(arr[i] == 0) {
                if(i == lastIdx - zeroCount) {
                    arr[lastIdx] = 0;
                    lastIdx--;
                    break;
                }
                zeroCount++;
            }
        }

        int r = lastIdx - zeroCount;

        for(int i = r; i >= 0; i--) {
            if(arr[i] == 0) {
                arr[i + zeroCount] = 0;
                zeroCount--;
                arr[i + zeroCount] = 0;
            }
            arr[i + zeroCount] = arr[i];
        }
    }
};

File: 1089-duplicate-zeros/1089-duplicate-zeros.js
/**
 * @param {number[]} arr
 * @return {void} Do not return anything, modify arr in-place instead.
 */
var duplicateZeros = function (arr) {
    let res = [];
    let count = 0;

    arr.forEach((ele) => {
        if (count < arr.length) {
            if (ele === 0) {
                res.push(0);
                count++;
            }
            if (count < arr.length) {
                res.push(ele);
            }
            count++;
        }
    })

    res.forEach((ele, idx) => arr[idx] = ele);
};